sql/sqlbase: prevent users from dropping any system tables

benesch opened the pull request:
See the commit messages and #13691 for details. In short: in existing clusters, users can drop non-config system tables (i.e., `system.ui`, `system.eventlog`, `system.rangelog` and `system.lease`). Doing so can break a cluster in weird ways, and the user can't even recreate the tables manually because they don't have the `CREATE` permission on the `system` database.

This PR revokes `CREATE` and `DROP` permissions on the aforementioned tables in new clusters. I'm not sure whether it's worth adding a migration to `REVOKE` those privileges in existing clusters, so I've left that out for now.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/13698)
<!-- Reviewable:end -->

benesch committed:
sql/sqlbase: clean up comments about system tables

Remove redundant (and incorrent) comments about `TestSystemTables`; that
test has been renamed to TestSystemTableLiterals and is already
mentioned at the top of the file. Additionally reorganize so it's clear
the top comment applies to all following `CREATE TABLE`s, not just the
system config tables.

benesch committed:
sql/sqlbase: prevent users from dropping any system tables

Cockroach prevents users from dropping system config tables, but allows
users to drop non-config system tables. This is problematic; the system
expects all system tables to exist, including non-config tables, and has
no facility to recreate them if the user drops them.

This commit ensures that, in new clusters, users cannot drop any system
tables. This commit does not attempt to revoke the CREATE and DROP
permissions on non-config system tables in existing clusters. Fixes
 #13691.

tamird commented:
Reviewed 1 of 1 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/sql/sqlbase/system.go, line 101 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13698#-KdXZ-QaOYbykTg1NuCg:-KdXZ-QaOYbykTg1NuCh:bre0d2m) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3178d91d18a3a8448eb4eaf32c86fb47164685f/pkg/sql/sqlbase/system.go#L101)):*
> ```Go
> );`
> 
> 	// blobs based on unique keys.
> ```

I wouldn't be upset if you rewrote or removed this comment.

---

*[pkg/sql/sqlbase/system.go, line 245 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13698#-KdXZIrgU-xWSNyGHHIx:-KdXZIrgU-xWSNyGHHIy:b80c1ms) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3178d91d18a3a8448eb4eaf32c86fb47164685f/pkg/sql/sqlbase/system.go#L245)):*
> ```Go
> 	}
> 
> 	// SystemConfigAllowedPrivileges describes the privileges allowed for each
> ```

nit: why move this?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13698)*
<!-- Sent from Reviewable.io -->

benesch commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/sql/sqlbase/system.go, line 245 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13698#-KdXZIrgU-xWSNyGHHIx:-KdX_-ZrPTL-jbIANnjP:bmjb1d6) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3178d91d18a3a8448eb4eaf32c86fb47164685f/pkg/sql/sqlbase/system.go#L245)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: why move this?
</blockquote></details>

The first `const` grouping (above) contains system config tables; the next (below) contains non-config tables. Moved it to its own section because it now contains permissions for both config and non-config tables (and has been renamed appropriately).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13698)*
<!-- Sent from Reviewable.io -->

benesch commented:
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/sql/sqlbase/system.go, line 101 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13698#-KdXZ-QaOYbykTg1NuCg:-KdXbiXXvMOuox3MjTJj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3178d91d18a3a8448eb4eaf32c86fb47164685f/pkg/sql/sqlbase/system.go#L101)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I wouldn't be upset if you rewrote or removed this comment.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13698)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r3, 3 of 3 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13698)*
<!-- Sent from Reviewable.io -->

mberhault commented:
Seems reasonable to me, but you may want to check with @mrtracy for the eventlog/ui tables, and @vivekmenezes for the lease table.
You should run the logic tests, some of them are failing due to this change.

benesch commented:
Yup, thanks! Should be fixed in the latest force-push.

tamird commented:
Reviewed 6 of 6 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13698)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented on pkg/sql/sqlbase/system.go:
```diff
@@ -31,8 +31,9 @@ import (
+// These system tables are part of the system config.
 const (
-	// NamespaceTableSchema is checked in TestSystemTables.
 	NamespaceTableSchema = `
```
given that these are exported const I'm surprised make check succeeds without a comment describe them.

vivekmenezes commented on pkg/sql/sqlbase/system.go:
```diff
@@ -439,3 +442,7 @@ func addSystemDatabaseToSchema(target *MetadataSchema) {
 func IsSystemConfigID(id ID) bool {
```
Is this still used?

benesch replied:
Yep, by `UPDATE`s, which seem to properly care whether the update is impacting the system config span or not.

vivekmenezes commented:
If an admin wants to blow away one of these tables they can always DELETE all rows. That's a good property that this PR retains. Thanks

benesch commented:
This needs a complete re-review, unfortunately—the original PR failed the upgrade tests, and for good reason. Table descriptor validation is aggressive about demanding the appropriate privileges for the root user on system tables, which means it was previously impossible to run two versions that expected different permissions in the same cluster. This PR now adds the capacity for a system table to have multiple allowed privileges for root. New clusters will get the new, proper privileges, but old clusters can continue to run with the old privileges.

Unfortunately there's no path to removing the old permissions until backwards-incompatible migrations are supported, but I still think this change is worthwhile in the meantime.

tamird commented:
Reviewed 6 of 6 files at r7.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/sql/sqlbase/privilege_test.go, line 202 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13698#-Kdltz7NVbOstOuZ2Ety:-Kdltz7NVbOstOuZ2Etz:b-2uuatw) ([raw file](https://github.com/cockroachdb/cockroach/blob/3684e4bc0667d0bc8681282c67a3e066e5e38720/pkg/sql/sqlbase/privilege_test.go#L202)):*
> ```Go
> 	id := ID(keys.MaxReservedDescID)
> 
> 	SystemAllowedPrivileges[id] = privilege.Lists{
> ```

can you check that there's not already a value there? just a sanity check.

---

*[pkg/sql/sqlbase/privilege_test.go, line 207 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13698#-Kdlu2FGQNI8xgUmyWG5:-Kdlu2FGQNI8xgUmyWG6:b-p6gsbx) ([raw file](https://github.com/cockroachdb/cockroach/blob/3684e4bc0667d0bc8681282c67a3e066e5e38720/pkg/sql/sqlbase/privilege_test.go#L207)):*
> ```Go
> 		{privilege.ALL},
> 	}
> 	defer func() {
> ```

you don't need the `func()`; `defer delete(.....)`

---

*[pkg/sql/sqlbase/privilege_test.go, line 211 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13698#-Kdlu7CyOhvkCCGTbI2p:-Kdlu7CyOhvkCCGTbI2q:b8w348k) ([raw file](https://github.com/cockroachdb/cockroach/blob/3684e4bc0667d0bc8681282c67a3e066e5e38720/pkg/sql/sqlbase/privilege_test.go#L211)):*
> ```Go
> 	}()
> 
> 	var descriptor *PrivilegeDescriptor
> ```

why the var? just use `:=` below.

if you want to separately scope each value of `descriptor` you can use a scope:

```
{
  descriptor := ...
  ...
}
// descriptor is not defined here
{
  descriptor := ...
  ...
}
```

---

*[pkg/sql/sqlbase/privilege_test.go, line 282 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13698#-KdluTGK6hBOz5Yighc8:-KdluTGK6hBOz5Yighc9:b-ye2grh) ([raw file](https://github.com/cockroachdb/cockroach/blob/3684e4bc0667d0bc8681282c67a3e066e5e38720/pkg/sql/sqlbase/privilege_test.go#L282)):*
> ```Go
> 	descriptor.Revoke(security.RootUser, privilege.List{privilege.SELECT})
> 	descriptor.Grant(security.RootUser, privilege.List{privilege.SELECT})
> 	if err := descriptor.Validate(id); err == nil {
> ```

check for the specific error you expect, either by direct equality if it's a sentinel error, or using `testutils.IsError`.

---

*[pkg/sql/sqlbase/system.go, line 37 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/13698#-KdXXUjJQiuDetDfeBZ8-r5-37:-Kdlsd73kKbWukUSR82S:bt7fffy) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7fbf8fcb03dfd7b943e6af71c23f414aa23bdae/pkg/sql/sqlbase/system.go#L37)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

given that these are exported const I'm surprised make check succeeds without a comment describe them.
</blockquote></details>

A single comment describing a "const" block is sufficient.

---

*[pkg/sql/sqlbase/system.go, line 139 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13698#-KdltQFVKXEX5jT4VM8I:-KdltQFVKXEX5jT4VM8J:bb6mt9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/3684e4bc0667d0bc8681282c67a3e066e5e38720/pkg/sql/sqlbase/system.go#L139)):*
> ```Go
> // reasons only!
> //
> // If we supported backwards-incompatible migrations, pruning allowed privileges
> ```

:+1:

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13698)*
<!-- Sent from Reviewable.io -->

benesch commented:
Review status: 6 of 7 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/sql/sqlbase/privilege_test.go, line 202 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13698#-Kdltz7NVbOstOuZ2Ety:-Kdm7vEmlsHZG4fybNcV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3684e4bc0667d0bc8681282c67a3e066e5e38720/pkg/sql/sqlbase/privilege_test.go#L202)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

can you check that there's not already a value there? just a sanity check.
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/privilege_test.go, line 207 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13698#-Kdlu2FGQNI8xgUmyWG5:-Kdm7vgPOQ_Zcnme450S:b2dkds0) ([raw file](https://github.com/cockroachdb/cockroach/blob/3684e4bc0667d0bc8681282c67a3e066e5e38720/pkg/sql/sqlbase/privilege_test.go#L207)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

you don't need the `func()`; `defer delete(.....)`
</blockquote></details>

Thanks :disappointed: 

---

*[pkg/sql/sqlbase/privilege_test.go, line 211 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13698#-Kdlu7CyOhvkCCGTbI2p:-Kdm86xixkBomLaisa9D:b-2byab1) ([raw file](https://github.com/cockroachdb/cockroach/blob/3684e4bc0667d0bc8681282c67a3e066e5e38720/pkg/sql/sqlbase/privilege_test.go#L211)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why the var? just use `:=` below.

if you want to separately scope each value of `descriptor` you can use a scope:

```
{
  descriptor := ...
  ...
}
// descriptor is not defined here
{
  descriptor := ...
  ...
}
```
</blockquote></details>

Yup, the asymmetry of the `:=` and `=` drives me crazy. Scopes work nicely for separating the flows in the test—thanks!

```
descriptor := thingy()
moreThingy()

descriptor = thingy2()
moreThingy2()
```

---

*[pkg/sql/sqlbase/privilege_test.go, line 282 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13698#-KdluTGK6hBOz5Yighc8:-Kdm8KHXzNqf_sFdOewz:b-dtkds3) ([raw file](https://github.com/cockroachdb/cockroach/blob/3684e4bc0667d0bc8681282c67a3e066e5e38720/pkg/sql/sqlbase/privilege_test.go#L282)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

check for the specific error you expect, either by direct equality if it's a sentinel error, or using `testutils.IsError`.
</blockquote></details>

Done, thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13698)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13698)*
<!-- Sent from Reviewable.io -->

benesch commented:
@mberhault would you mind to take another look at this? This rewrite touches a bunch of your permissions code.

benesch committed:
sql: produce correct diff output in SystemTableLiterals test

TestSystemTableLiterals compares two protobufs for equality. If the two
protobufs are not equal, it attempts to print out where they differ.
Unfortunately, the diff algorithm would report different pointers to the
same data (which proto.Equal considers equivalent) as the source of the
difference when the real error could be hidden in later fields. This
commit switches the test to use a proper deep-diff printer bundled with
a pretty printing library we already vendor to avoid confusing future
young engineers.

tamird commented:
Reviewed 1 of 1 files at r9.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/system_table_test.go, line 127 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/13698#-Ke5pMrA8ju433oxVQvO:-Ke5pMrA8ju433oxVQvP:b-8k48av) ([raw file](https://github.com/cockroachdb/cockroach/blob/7365727c49f570f5c07bfb99f9ddf045d73385dc/pkg/sql/system_table_test.go#L127)):*
> ```Go
> 		}
> 		if !proto.Equal(&test.pkg, &gen) {
> 			pretty.Ldiff(t, &test.pkg, &gen)
> ```

I don't think we use this function anywhere. Instead, can you `%s` the value of `pretty.Diff` into the message below?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13698)*
<!-- Sent from Reviewable.io -->

benesch commented:
*[pkg/sql/system_table_test.go, line 127 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/13698#-Ke5pMrA8ju433oxVQvO:-Ke6pcAFAojBXOihBtWJ:byefy1e) ([raw file](https://github.com/cockroachdb/cockroach/blob/7365727c49f570f5c07bfb99f9ddf045d73385dc/pkg/sql/system_table_test.go#L127)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I don't think we use this function anywhere. Instead, can you `%s` the value of `pretty.Diff` into the message below?
</blockquote></details>

What's the benefit of that?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13698)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/system_table_test.go, line 127 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/13698#-Ke5pMrA8ju433oxVQvO:-KeOLzf9KtBTc2joazzQ:bsytboy) ([raw file](https://github.com/cockroachdb/cockroach/blob/7365727c49f570f5c07bfb99f9ddf045d73385dc/pkg/sql/system_table_test.go#L127)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

What's the benefit of that?
</blockquote></details>

It's consistent with existing usage.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13698)*
<!-- Sent from Reviewable.io -->

benesch commented:
Review status: 0 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/system_table_test.go, line 127 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/13698#-Ke5pMrA8ju433oxVQvO:-Ke_7HfYsCEpC2n4f0UL:b-o9yyyz) ([raw file](https://github.com/cockroachdb/cockroach/blob/7365727c49f570f5c07bfb99f9ddf045d73385dc/pkg/sql/system_table_test.go#L127)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

It's consistent with existing usage.
</blockquote></details>

Ok, switched it over. FWIW every existing use that doesn't `strings.Join(pretty.Diff(...), "\n")`is going to be unreadable garbage if the diff is large.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13698)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 6 of 7 files at r11, 1 of 1 files at r12.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/sql/system_table_test.go, line 127 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/13698#-Ke5pMrA8ju433oxVQvO:-Ke_8Av5_Si7096k-1fi:bxdv0ac) ([raw file](https://github.com/cockroachdb/cockroach/blob/7365727c49f570f5c07bfb99f9ddf045d73385dc/pkg/sql/system_table_test.go#L127)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Ok, switched it over. FWIW every existing use that doesn't `strings.Join(pretty.Diff(...), "\n")`is going to be unreadable garbage if the diff is large.
</blockquote></details>

Interesting. Feel free to audit existing usage in a different PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13698)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13698#-:-KedtG3TxlkqUYcfL6Kq:bnfp4nl)*
<!-- Sent from Reviewable.io -->

benesch commented:
@mrtracy agrees system.ui doesn't need `CREATE, DROP` privs, so merging!
